### PR TITLE
examples: Moved CI infos to Makefile.ci

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

- Decluttered the Makefiles by moving the `BOARD_INSUFFICIENT_MEMORY` lists from each `Makefile` to a `Makefile.ci`
- Optimized the list for use of tools:
    - One entry per line reduces the number of merge conflicts
    - One entry per line allows alphabetical sorting e.g. via Vim's sort command
- Sorted all lists alphabetically

### Testing procedure

- Check the diffs
- Let the CI do its thing

### Issues/PRs references

https://github.com/RIOT-OS/RIOT/issues/11128, https://github.com/RIOT-OS/RIOT/issues/9965

### CODE REVIEW

**Comment by maribu** [comment]
If this PR gets swiftly merged, I might be able to do the same cleanup for the `Makefile`s in `tests` today ;-)

**Comment by benpicco** [comment]
I like the idea!
But instead of having `include Makefile.ci` in each Makefile, how about adding `-include Makefile.ci` to [`Makefile.include`](https://github.com/RIOT-OS/RIOT/blob/master/Makefile.include#L22)

**Comment by miri64** [comment]
I'm not sure, what this is trying to accomplish... The clutter is just moved to a different spot. Also: the lists are not just about CI, but especially for BLACKLIST/WHITELIST to warn a user, that the app might not be applicable for the board they are using. So the name might be misplaced.

**Comment by miri64** [comment]
> If this PR gets swiftly merged, I might be able to do the same cleanup for the `Makefile`s in `tests` today ;-)

A big build system change like this shouldn't be merged right before feature freeze oO.

**Comment by benpicco** [comment]
> I'm not sure, what this is trying to accomplish... The clutter is just moved to a different spot.

Well the clutter is removed from the main Makefile which is actually read by humans, especially the examples.

With the clutter contained to one file, it can clutter that file up as much as it lies - a 100 line Makefile would be atrocious to read, but when it's contained to one file that *only* does that, the list can take up as much space as it wants, making merging easier.

**Comment by cladmi** [comment]
For the definition change at least, all values are still the same here.

I used this to print the variables for all values, then sorted to be able to compare and the sorted values are the same as the reference commit.

```
for app in $(make info-applications); do make --no-print-directory -C ${app} info-debug-variable-APPLICATION info-debug-variable-BOARD_WHITELIST info-debug-variable-BOARD_BLACKLIST info-debug-variable-BOARD_INSUFFICIENT_MEMORY; done | awk '{ delete a; for (i=1; i<=NF; i++) a[$i]++; n=asorti(a, b); for (i=1; i<=n; i++) printf b[i]" "; print "" }'
```

https://stackoverflow.com/questions/8802734/sorting-and-removing-duplicate-words-in-a-line/25823667#25823667

`BOARD_BLACKLIST` and `BOARD_WHITELIST` are not related to `CI`.
They are currently compensating for a lack of required features and thus do not belong to a `Makefile.ci`.


> > A big build system change like this shouldn't be merged right before feature freeze oO.
> 
> This PR only moves lists from one file to another and sorts their content. As the order of the entries in the list is of no semantic meaning, this PR doesn't change any semantics. But I wouldn't mind waiting a bit either.

This is a conceptual change that these definitions must be moved to another entity to define them.

The line based definition and re-ordering is not and could go before.


When doing such a change, it is also good to add some static checks to prevent having the old pattern re-appearing again. Many existing PRs do not follow this scheme and will be merged without being updated.

**Comment by miri64** [comment]
> A big build system change like this shouldn't be merged right before feature freeze oO.

Big in the sense of lines touched ;-). I know that the impact shouldn't be that high.

**Comment by miri64** [comment]
> Also: the lists are not just about CI, but especially for BLACKLIST/WHITELIST to warn a user, that the app might not be applicable for the board they are using. So the name might be misplaced.

What about this point though?

**Comment by maribu** [comment]
> > Also: the lists are not just about CI, but especially for BLACKLIST/WHITELIST to warn a user, that the app might not be applicable for the board they are using. So the name might be misplaced.
> 
> What about this point though?

How about renaming `Makefile.ci` to no longer indicate a strong connection to the CI? Something like `Makefile.boardlists`?

**Comment by maribu** [comment]
> Big in the sense of lines touched ;-)

True. So lets wait with merging than. I could also the time add a second commit to apply the approach to tests in this PR, once the discussion reached an agreement.

**Comment by benpicco** [comment]
> `BOARD_BLACKLIST` and `BOARD_WHITELIST` are not related to `CI`.
> They are currently compensating for a lack of required features and thus do not belong to a `Makefile.ci`.

They could go to `Makefile.local` then.

**Comment by maribu** [comment]
I think the name `Makefile.local` is already used to allow developers to add configurations locally to their system that is not intended to be upstreamed. But I'm not sure if I remember this correctly. I think @cladmi is heavily using the features of locally extending the build system to directly select the correct programmer by the value in the `BOARD` variable on his testing rig, right?

**Comment by maribu** [comment]
> at least for the `BOARDS_WHITELIST` and `BOARDS_BLACKLIST` there is IMHO a better way to improve by using features, see #12423 as an RFC example.

I agree that using features is generally a better idea than using the black and the whitelist. Even if some of the features would have to be maintained for every board individually, this would be already hugely less pain - the boards are pretty static while the applications are moving targets. But I honestly don't think that this could be done in a single PR, but rather quite a bunch.

I'd say moving those lists out of the main Makefile and having the lists alphabetically sorted and one board per line is a simple and swift way to significantly lessen the pain those lists cause. And this in no way conflicts with any approaches that reduce the number of cases those lists have to be used. I think it is the opposite: Once the lists are easier to manage with scripts (due to having one board in a row and in "lists-only-files"), it will less effort to clean up those list when features can be used instead.

**Comment by smlng** [comment]
@maribu I somewhat agree for the `BOARDS_INSUFFICIENT_MEMORY` list, that moving this into a separate file makes it easier to update the list (also allows to copy this to other apps and tests where the same applies). 

For black and white list I would prefer to use the feature mechanism as proposed in #12423.

**Comment by maribu** [comment]
> For black and white list I would prefer to use the feature mechanism as proposed in #12423.

Did you notice that I 100% agree on that? ;-)

I only think that until this goal is reached, having those lists separated out, sorted, and formatted one board a line is beneficial.

**Comment by smlng** [comment]
> Did you notice that I 100% agree on that? ;-)

I did ... my suggestion would be to move in parallel, i.e. moving `BOARDS_INSUFFICIENT_MEMORY` in a separate Makefile (your PR here) and remove black/whitelists by using features (either extend #12423 or add PRs per feature). Instead of moving the latter too and changing it later again.

**Comment by maribu** [comment]
> I did ... my suggestion would be to move in parallel, i.e. moving `BOARDS_INSUFFICIENT_MEMORY` in a separate Makefile (your PR here) and remove black/whitelists by using features (either extend #12423 or add PRs per feature). Instead of moving the latter too and changing it later again.

OK. Let's do this. @miri64: This way also the name `Makefile.ci` becomes reasonable, right?

@kaspar030: https://github.com/RIOT-OS/RIOT/pull/9081 is not concerned about `BOARD_INSUFFICIENT_MEMORY`, so that way the order in which https://github.com/RIOT-OS/RIOT/pull/9081 and this PR are merged no longer matters, right?

**Comment by kaspar030** [comment]
> @kaspar030: #9081 is not concerned about `BOARD_INSUFFICIENT_MEMORY`, so that way the order in which #9081 and this PR are merged no longer matters, right?

correct.

**Comment by kaspar030** [comment]
Why don't we solve BOARD_INSUFFICIENT_MEMORY so it doesn't pose *any* maintanance anymore? Because, all we're doing at the moment is waiting for CI output and if linking fails, we add the board to the list. I'm sure we can, in the CI case, *always* link and find a way to recognize failures due to memory constraints, and handle them differently.

**Comment by smlng** [comment]
@kaspar030 that would be best, and the idea came up several times, but never got implemented due to constraints. IIRC one argument was to optimise build times by ignoring applications in the CI that are known to fail linking anyway.

**Comment by smlng** [comment]
I would go for this, bc its basically only restructuring existing build-system config. Hence should not break anything, while still enhancing readability of Makefiles. 

Getting *build* and *try-linking* working with the CI will probably take longer.

**Comment by maribu** [comment]
May I squash?

**Comment by smlng** [comment]
Testing via Murdock should suffice for everything in the RIOT tree.

**Comment by kaspar030** [comment]
> May I squash?

Hm, I still don't like adding >300 files each just containing the insufficient memory list. If you think expanding the list to one line per board, can we maybe do that first and postpone the Makefile.ci discussion?

**Comment by smlng** [comment]
> Hm, I still don't like

IIRC this is not a valid argument.

**Comment by maribu** [comment]
> Hm, I still don't like adding >300 files each just containing the insufficient memory list.

Pro separate that out:
- The chance of people opening `Makefile` wanting to read/edit `BOARDS_INSUFFICIENT_MEMORY` is very low, so having it in `Makefile` will waste their time scrolling over it
- Especially in the `examples` those list will only confuse new users of RIOT, who really do not need to understand the background of `BOARDS_INSUFFICIENT_MEMORY` when just starting to work with RIOT
- Scripts using `find RIOT -type f -name Makefile.ci -exec <FOO> {} \;` are much less likely to cause collateral damage than `find RIOT -type f -name Makefile -exec <FOO> {} \;`
- If there is a merge conflict with a `Makefile` right now, it is like 99% of the time due to `BOARDS_INSUFFICIENT_MEMORY` being touched. The chance of accidentally breaking something imported is much lower when the merge conflict is contained in a `Makefile.ci` containing only the `BOARDS_INSUFFICIENT_MEMORY`

Against splitting it out
- @kaspar030 doesn't like it

**Comment by kaspar030** [comment]
>     * @kaspar030 doesn't like it

I don't have more arguments other than even greater dispersion of our build configuration. I hoped that at least for applications we could get away with having one file...

**Comment by maribu** [comment]
@kaspar030: Consider this as the first step of removing the `BOARDS_INSUFFICIENT_MEMORY` list. With the current state, every `Makefile` would need to be touched. With this merged, `find RIOT -type f -name Makefile.ci -exec rm {} \;` will be sufficient, when the array is no longer needed ;-)

**Comment by smlng** [comment]
also please squash current changes, append check as commented above if possible; and finally: trigger CI (after squash).

**Comment by maribu** [comment]
> Open question/suggestion (referring to @cladmi comment): can we somehow check (and thus enforce) that `BOARDS_INSUFFICIENT_MEMORY` is always defined in `Makefile.ci` and not in `Makefile`?

I like the idea. But wouldn't it be required to get to that state first, before adding the check? (Right now only `examples` are handled, the bigger chunk will be the `tests`.)

Should I tackle the `tests` here as second commit, (and than the test could also be added), or should that be a new PR?

**Comment by smlng** [comment]
@maribu you could add the tests in https://github.com/RIOT-OS/RIOT/blob/master/dist/tools/buildsystem_sanity_check/check.sh and first only check path=`examples/*/Makefiles` to not specify `BOARDS_INSUFFICIENT_MEMORY` and later we extend this to `tests/*/Makefile` (or all Makefiles in general).

**Comment by miri64** [comment]
Another way of reducing those lists and the maintenance burden could be to list CPU_INSUFFICIENT_MEMORY instead of BOARD_INSUFFICIENT_MEMORY , since the memory constraints are less of a board constraint, but a CPU constraint, so listing the CPUs instead makes much more sense. The clearest advantage is that new boards of already existing CPUs don't need to be added in the boards initial import, the disadvantage is that it might not be as clear with the CI what to add as boards are.

**Comment by maribu** [comment]
> Another way of reducing those lists and the maintenance burden could be to list CPU_INSUFFICIENT_MEMORY instead of BOARD_INSUFFICIENT_MEMORY

Nice idea. However, board specific bootloaders are not taking into account. As a result the board with the biggest bootloader will force the decision upon other boards. Also, board specific stuff like number of network interfaces used can increase requirements for some board.

Now having written this, I don't think anymore that this will work :-/

**Comment by maribu** [comment]
@smlng: Check has been added as requested. It works as expected *on my machine*, hopefully also on others :-)

**Comment by maribu** [comment]
Aha: The check also works on travis: https://travis-ci.org/RIOT-OS/RIOT/builds/596637735?utm_source=github_status&utm_medium=notification

I'll rebase to also apply the change for `examples/suit_update`

**Comment by maribu** [comment]
I believe the issue [Murdock points out](https://ci.riot-os.org/RIOT-OS/RIOT/12406/161bb94af1b0db0bfc0ab865b36c920a6f78ad48/output/makefile_broken.txt) is a bug in the CI:

```
/bin/sh: 1: examples/suit_update: Permission denied
```

Can someone help? I triggered Murdock now three times. In the first run, every test failed with `/bin/sh: 1: ./.murdock: not found`. The last two runs showed the message above (1 fail, all other tests pass). I think something about this PR triggers bugs in the CI...

**Comment by miri64** [comment]
For me, bugging ci@riot-labs.de works most of the time ;-).

File: examples/suit_update/Makefile.ci [hunk 1]
```diff
@@ -0,0 +1,30 @@
+BOARD_INSUFFICIENT_MEMORY :=
```
Comment (kaspar030) [inline examples/suit_update/Makefile.ci]: missing backslash

Comment (kaspar030) [inline examples/suit_update/Makefile.ci]:
```suggestion
BOARD_INSUFFICIENT_MEMORY := \
```

Comment (maribu) [reply examples/suit_update/Makefile.ci]: Indeed. Sorry for making the noise!

**Comment by maribu** [comment]
May I squash?

**Comment by maribu** [comment]
@smlng: All green :-)

**Pull request merged**